endpoints/web.py

```diff
@@ -897,3 +898,82 @@ def redirect_to_namespace(namespace):
         return redirect(url_for("web.org_view", path=namespace))
     else:
         return redirect(url_for("web.user_view", path=namespace))
+
+
+def has_users():
+    """
+    Return false if no users in database yet
+    """
+    return bool(list(User.select().limit(1)))
+
+
+@web.route("/api/v1/user/initialize", methods=["POST"])
+def user_initialize():
+    """
+    Create initial user in an empty database
+    """
+
+    # Ensure that we are using database auth.
+    if not features.USER_INITIALIZE:
+        response = jsonify({"message": "Cannot initialize user, FEATURE_USER_INITIALIZE is False"})
+        response.status_code = 400
+        return response
+
+    # Ensure that we are using database auth.
+    if app.config["AUTHENTICATION_TYPE"] != "Database":
+        response = jsonify({"message": "Cannot initialize user in a non-database auth system"})
+        response.status_code = 400
+        return response
+
+    if has_users():
+        response = jsonify({"message": "Cannot initialize user in a non-empty database"})
+        response.status_code = 400
+        return response
+
+    user_data = request.get_json()
+    try:
+        prompts = model.user.get_default_user_prompts(features)
+        new_user = model.user.create_user(
+            user_data["username"],
+            user_data["password"],
+            user_data.get("email"),
+            auto_verify=True,
+            email_required=features.MAILING,
+            is_possible_abuser=False,
+            prompts=prompts,
+        )
+        success, headers = common_login(new_user.uuid)
+        if not success:
+            response = jsonify({"message": "Could not login. Failed to initialize user"})
+            response.status_code = 400
+            return response
+
+        result = {
+            "username": user_data["username"],
+            "email": user_data.get("email"),
+            "encrypted_password": authentication.encrypt_user_password(
+                user_data["password"]
+            ).decode("ascii"),
+        }
+
+        if user_data.get("access_token"):
+            model.oauth.create_application(
+                new_user,
+                "automation",
+                "",
+                "",
+                client_id=user_data["username"],
+                description="Application token generated via /api/v1/user/initialize",
+            )
+            token_string = random_string_generator(length=40)()
+            scope = "org:admin repo:admin repo:create repo:read repo:write super:user user:admin user:read"
+            model.oauth.create_user_access_token(
+                new_user, user_data["username"], scope, access_token=token_string
+            )
+            result["access_token"] = token_string
+
+        return (result, 200, headers)
+    except model.user.DataModelException as ex:
+        response = jsonify({"message": "Failed to initialize user: " + str(ex)})
+        response.status_code = 400
+        return response
```

Comment:
Wasn't this previously in the `user.py` or `superuser.py` module? Why did you move it to `web`?

Reply:
This in web.py due to the csrf and api needing to be unauthenticated. We spent a long time back and forth on this.